### ISSUE TITLE
# [bug] Qwen3.5 packing under sp=1 silently leaks across samples (GDN not isolated) on transformers 5.2.0~5.8.1

### ISSUE THREAD

**Issue by realgump** [modelscope/ms-swift#9618]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new bug report. / 我已经搜索过现有的 issues，确认这是一个新的 bug report。

## Summary

Under `--packing true` (or `--padding_free true`) with `--sequence_parallel_size 1` (the default), Qwen3.5's GatedDeltaNet (GDN) linear-attention layers do **not** isolate packed sample boundaries. The causal-conv / recurrent state leaks across samples **silently** — no error, no warning, training loss may look normal but model quality degrades. This is exactly the failure mode described in huggingface/transformers#46093.

`sp>=2` is unaffected (the GDN self-derives `cu_seqlens` from `real_position_ids`, bypassing the broken path).

PR #9571 ([transformers] support qwen3_5 padding_free/packing) intended to support sp=1 + packing — it made `Qwen3_5GatedDeltaNet.forward` read `cu_seq_lens_q` from kwargs — but on the transformers versions swift currently declares support for (`>=5.2.0`), that path never fires, because `cu_seq_lens_q` is dropped one hop before reaching the GDN.

## Affected versions

- swift: current `main` (contains PR #9571)
- transformers: **5.2.0 ~ 5.8.1** (the range swift declares via `requires=['transformers>=5.2.0']`)
- transformers **5.9.0+ is already fixed upstream** (see "Upstream status" below)

## Root cause

The collator (`Qwen2VLTemplate._data_collator`, inherited by `Qwen3_5Template`) correctly produces `cu_seq_lens_q`, and it is forwarded down via `**kwargs` through `ForConditionalGeneration → Qwen3_5Model → Qwen3_5TextModel → Qwen3_5DecoderLayer`.

But transformers 5.2.0's `Qwen3_5DecoderLayer.forward` calls `self.linear_attn(...)` in the `linear_attention` branch with only 4 named args and does **NOT** forward `**kwargs` (only the `full_attention` branch does):

```python
# transformers 5.2.0, Qwen3_5DecoderLayer.forward
if self.layer_type == "linear_attention":
    hidden_states = self.linear_attn(
        hidden_states=hidden_states,
        cache_params=past_key_values,
        cache_position=cache_position,
        attention_mask=attention_mask,
    )                          # <- **kwargs NOT forwarded, cu_seq_lens_q dropped here
elif self.layer_type == "full_attention":
    hidden_states, _ = self.self_attn(..., **kwargs)   # only full_attention forwards kwargs
```

So `cu_seq_lens_q` is dropped right before reaching the GDN. PR #9571's sp=1 branch:

```python
# swift/model/models/qwen.py, _run_qwen3_5_gated_delta_net_sequence_parallel_forward
if sp_enabled:
    cu_seqlens = _get_qwen3_5_cu_seqlens_q()
else:
    cu_seqlens = kwargs.get('cu_seq_lens_q')   # <- always None on transformers 5.2.0~5.8.1
...
if not sequence_parallel.enabled() and 'cu_seq_lens_q' not in kwargs:
    return origin_forward(...)                  # <- GDN takes this branch: no isolation
```

`kwargs.get('cu_seq_lens_q')` is always `None`, the GDN falls back to `origin_forward` (no boundary isolation) → cross-sample leak.

Data flow (verified on transformers 5.2.0 + Qwen3.5-4B):

```
collator produces cu_seq_lens_q = tensor([0, L_a, L_a+L_b])   ✓
  -> ForConditionalGeneration.forward(**kwargs)                ✓ forwards
  -> Qwen3_5Model.forward -> language_model(**kwargs)          ✓ forwards
  -> Qwen3_5TextModel.forward -> decoder_layer(**kwargs)       ✓ forwards
  -> Qwen3_5DecoderLayer.forward(**kwargs)
       linear_attention branch: self.linear_attn(4 args)       ✗ drops cu_seq_lens_q
  -> Qwen3_5GatedDeltaNet.sp_linear_forward: kwargs has NO cu_seq_lens_q
       -> 'cu_seq_lens_q' not in kwargs  =>  origin_forward    ✗ no isolation => LEAK
```

## Reproduction

Verified end-to-end with a real collator + GDN forward probe (Qwen3.5-4B, transformers 5.2.0, sp=1 + packing):

| Metric | Value |
|---|---|
| collator produces `cu_seq_lens_q` | ✅ `tensor([0, L_a, L_a+L_b])` |
| GDN receives `cu_seq_lens_q` | ❌ False (dropped at DecoderLayer) |
| GDN branch taken | `origin_forward` (no isolation) |
| packed B-region vs separate B, first GDN layer `max\|Δ\|` | **0.068** (leak, ~39% of A-vs-B magnitude) |

Reproduction snippet:

```python
from swift.template import get_template
template = get_template(tokenizer, template_type='qwen3_5', padding_free=True, sequence_parallel_size=1)
batch = template._data_collator([enc_a, enc_b])
# batch['cu_seq_lens_q'] == tensor([0, L_a, L_a+L_b])   # produced ✓
# but GDN forward kwargs has NO cu_seq_lens_q -> origin_forward -> trailing sample B leaks
```

Notes:
- The leak is on the **trailing** sample (B): its conv/recurrent state is polluted by the leading sample (A). The leading sample is unaffected (causal conv starts from zero initial state), so the leak is easy to miss in a casual smoke test.
- Standard self-attention layers are isolated correctly (they rely on boundary-reset `position_ids` + FA varlen); only the GDN linear-attention layers leak.

## Upstream status (transformers already fixed this in 5.9.0)

The root cause was fixed **upstream in transformers 5.9.0** (2026-05-20), in the same way a swift-side patch would:

```python
# transformers 5.9.0+, Qwen3_5DecoderLayer.forward
if self.layer_type == "linear_attention":
    hidden_states = self.linear_attn(
        hidden_states=hidden_states,
        cache_params=past_key_values,
        attention_mask=attention_mask,
        **kwargs,                       # <- added in 5.9.0
    )
# Qwen3_5GatedDeltaNet.forward now has **kwargs and uses:
#   causal_conv1d_fn(..., seq_idx=kwargs.get("seq_idx"))
#   chunk_gated_delta_rule(..., cu_seqlens=kwargs.get("cu_seq_lens_q"))
```

Verified by checking each release's `modeling_qwen3_5.py`:
- 5.2.0 ~ 5.8.1: `linear_attention` branch does **not** forward `**kwargs` (bug present)
- 5.9.0 ~ 5.12.1 (latest): forwards `**kwargs` (fixed)

## Suggested resolution (maintainer decision)

Two options, depending on whether swift still wants to support transformers 5.2.0~5.8.1:

1. **Bump the requirement**: change `requires=['transformers>=5.2.0']` to `>=5.9.0` for Qwen3.5 models (and document the leak for users who can't upgrade). Simplest; no swift code change.
2. **Backfill a swift-side patch** for 5.2.0~5.8.1: monkeypatch `Qwen3_5DecoderLayer.forward` to forward `cu_seq_lens_q` to `self.linear_attn` in the `linear_attention` branch (mirroring what transformers 5.9.0 does natively). Harmless on `>=5.9.0` (only intervenes when `cu_seq_lens_q` is present, otherwise delegates to the original forward). I have a working patch (+61 lines, single file `swift/model/models/qwen.py`) verified to fix the leak (`max|Δ|` 0.068 → 0.000000) — happy to open a PR if this direction is preferred.

## Related

- PR #9571 ([transformers] support qwen3_5 padding_free/packing) — intended sp=1 support, but ineffective on transformers 5.2.0~5.8.1 due to this bug
- huggingface/transformers#46093 (packing on Qwen3.5 / hybrid linear attention — silent cross-sample corruption)
- huggingface/transformers#44717 (related padding-free discussion)

### How to Reproduce / 如何复现

...

**Comment by Jintao-Huang** [comment]
fixed
